### ISSUE TITLE
Poor type inference for `reduce`

### ISSUE THREAD

**Issue by andy-ms** [Microsoft/TypeScript#25454]
**TypeScript Version:**  3.0.0-dev.20180705

**Code**

```ts
function toStrings(arr: ReadonlyArray<object>): string[] {
	return arr.reduce((acc, obj) => {
		acc.push(obj.toString());
		return acc;
	}, [] as string[]);
}
```

**Expected behavior:**

No error.

**Actual behavior:**

```
src/a.ts:2:2 - error TS2322: Type 'object' is not assignable to type 'string[]'.
  Property 'length' is missing in type '{}'.

2  return arr.reduce((acc, obj) => {
   ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
3  	acc.push(obj.toString());
  ~~~~~~~~~~~~~~~~~~~~~~~~~~~
4  	return acc;
  ~~~~~~~~~~~~~
5  }, [] as string[]);
  ~~~~~~~~~~~~~~~~~~~~

src/a.ts:3:7 - error TS2339: Property 'push' does not exist on type 'object'.

3  	acc.push(obj.toString());
```

No error if I explicitly specify `arr.reduce<string[]>`.
No error if I remove the first two overloads to `reduce`, which are non-generic.

**Comment by willyboy** [comment]
I think this goes here:
```
const things: Things[] = [{num:1}, {num:5}, {num: 2}];
this.things.slice(0, index).reduce((prevVal, curVal): number => {
      return prevVal.num + curVal.num;
    });
```

Error:
_Argument of type '(prevVal: Things, curVal: Things) => number' is not assignable to parameter of type '(previousValue: Things, currentValue: Things, currentIndex: number, array: Things[]) => Scene'._

Expected result is no error. We should be allowed to reduce an object to a single value that isn't the same object.

**Comment by lolmaus** [comment]
@willyboy It's indicating a bug in your code.

You're not passing an intial value into `.reduce()`, so JavaScript expects both `prevVal` and `curVal` to be `Things`. But you're returning a number from the callback, which goes into `prevVal`.

TypeScript says that you have `=> number` but it should be `=> Things`. This is a legit, helpful error message.

Check out [this snippet](https://www.typescriptlang.org/play?#code/JYOwLgpgTgZghgYwgAgCoAtQHNkG8CwAUMsiAK4C2AXKZQEbQDcRAvkUQgPYgDOYyYTCCw8aGbAG0AusgC8yCbnLUAjCwA0eZVQCsGrZRoAmFlOaEig7DwB0UCABMySABQuADvYBuANTgAbTQQyKD9-AEoaZQYoOQA+PCISezAQkGRPCF8Am2VkAGpkYNCc5XMWcPNLIRE7R2cIN3seMn8wKPpoTSthMRrI2goY+MTiZBS08YgWtoKBGtzKcs0ABkqiIA) for comparison.

**Comment by russoedu** [comment]
I can add something to this.

I have an array with `zeroes` and `ones` created from a comparison (the example I'm posting here is a stupid version of it, just so the array type is (0 | 1)[]).

The `reduce` result is a positive number, but TS infers it will be 0 | 1.

```
const arr = [0, 1, 0, 0, 1, 1, 1, 1, 1, 1].map(v => v === 0 ? 0 : 1)
const sum = arr.reduce((p, c) => p + c, 0)
```

![image](https://user-images.githubusercontent.com/472146/172556816-283c1dc1-98d1-4cb9-9d8f-9e87ec9c36d2.png)

**Comment by Vulcagon** [comment]
I have a similar issue.

```TypeScript
const recordArray: Record<string, any>[] = [
    { 'a': 1, 'b': 'B', 'c': [1, 2, 3] },
    { 'd': 4, 'e': 'E', 'f': [4, 5, 6] },
    { 'g': 7, 'h': 'H', 'i': [7, 8, 9] }
];

const reducedArray = recordArray.reduce(
    (map, record) => map,
    new Map<string, string>()
);
```

In this code `reducedArray` and the `map` parameter inside the `reduce` function are being inferred of type `Record<string, any>` instead of `Map<string, string>`. This also happens if the accumulator is an array or a `Set` and probably with many other types as well, but it works with simple types like `string` or `number`.

The example above is a pretty stupid example but I think it shows the issue pretty clear. My real code looks more like the following

```TypeScript
const reducedArray = recordArray.reduce(
    (map, record) => condition
        ? map.set(...)
        : map,
    new Map<...>()
);
```
